README.md
# Ardens

Ardens is an [Arduboy](https://www.arduboy.com/) simulator that is useful for profiling and debugging.

![screenshot](img/screenshot.png)

[Try out the full debugger here.](https://tiberiusbrown.github.io/Ardens/)
Drag and drop an [Arduboy game](https://community.arduboy.com/c/games/35) (.hex/.elf/.arduboy) and any FX data (.bin) onto the page.

Try the minimal player with one of the game URLs below:

- [ArduChess](https://tiberiusbrown.github.io/Ardens/player.html?file=https://raw.githubusercontent.com/tiberiusbrown/arduchess/master/arduchess/arduchess.hex)
- [ArduGolf](https://tiberiusbrown.github.io/Ardens/player.html?file=https://raw.githubusercontent.com/tiberiusbrown/arduboy_minigolf/master/ardugolf.hex)
- [ArduRogue](https://tiberiusbrown.github.io/Ardens/player.html?file=https://raw.githubusercontent.com/tiberiusbrown/ardurogue/master/ardurogue.hex)

## Keyboard Bindings

|   Key  | Purpose                            |
|:------:|------------------------------------|
| Arrows | Arduboy: directional buttons       |
|    A   | Arduboy: A button                  |
|   S/B  | Arduboy: B button                  |
|   F2   | Save screenshot                    |
|   F3   | Toggle GIF recording               |
|   F4   | (Debugger only) Save snapshot      |
|   F5   | (Debugger only) Pause/Continue     |
|   F11  | Toggle fullscreen                  |
|    O   | (Debugger only) Settings window    |
|    P   | (Debugger only) Toggle player mode |
|    R   | Rotate display                     |

## Creating a link to a playable Arduboy game

There is a web build hosted on GitHub pages: to create a link to a playable Arduboy game using Ardens Player, you can follow these patterns:

`https://tiberiusbrown.github.io/Ardens/player.html?file=https://example.com/game.arduboy`
`https://tiberiusbrown.github.io/Ardens/player.html?file=https://example.com/game.hex&file=https://example.com/game.bin`

The web builds of the debugger and player accept the following additional URL parameters (e.g., add `&param1=value1&param2=value2` to the URL).

#### `g` or `grid`
Overlays the display with a grid to allow easier disambiguation of adjacent pixels.
Values: `none`, `normal`, `red`, `green`, `blue`, `cyan`, `magenta`, `yellow`, `white`

#### `p` or `palette`
Adjusts the color palette of the display.
Values: `default`, `retro`, `lowcontrast`, `highcontrast`

#### `af` or `autofilter`
If enabled, applies temporal filtering to the display to help properly render grayscale games.
Values: `0`, `1`

#### `f` or `filter`
(Debugger only) Conditionally applies an upscaling filter.
Values: `none`, `scale2x`, `scale3x`, `scale4x`, `hq2x`, `hq3x`, `hq4x`

#### `ds` or `downsample`
(Debugger only) Downsamples by an integer ratio after applying an upsample filter.
Values: `0`, `1`, `2`, `3`

#### `ori` or `orientation`
Rotates the display in 90 degree increments (directional button mapping is also rotated as if playing on a rotated Arduboy).
Values:
- `0` or `normal`
- `90` or `cw` or `cw90`
- `180` or `flip`
- `270` or `ccw` or `ccw90`

## Libraries Used

- [SDL](https://github.com/libsdl-org/SDL) (zlib)
- [sokol](https://github.com/floooh/sokol) (zlib)
- [LLVM Project](https://github.com/llvm/llvm-project) (Apache v2.0 with LLVM Exceptions) 
- [Dear ImGui](https://github.com/ocornut/imgui) (MIT)
- [ImPlot](https://github.com/epezent/implot) (MIT)
- [imgui_markdown](https://github.com/juliettef/imgui_markdown) (zlib)
- [ImGuiColorTextEdit](https://github.com/BalazsJako/ImGuiColorTextEdit) (MIT)
- [Simple-FFT](https://github.com/d1vanov/Simple-FFT) (MIT)
- [brunexgeek/hqx](https://github.com/brunexgeek/hqx) (Apache v2.0)
- [{fmt}](https://github.com/fmtlib/fmt) (MIT)
- [rapidjson](https://github.com/Tencent/rapidjson) (MIT)
- [Miniz](https://github.com/richgel999/miniz) (MIT)
- [Bitsery](https://github.com/fraillt/bitsery) (MIT)
- [stb_image_write](https://github.com/nothings/stb/blob/master/stb_image_write.h) (Public Domain)
- [GIF encoder](https://github.com/lecram/gifenc) (Public Domain)
- [AudioFile](https://github.com/adamstark/AudioFile) (MIT)
- [Emscripten Browser File Library](https://github.com/Armchair-Software/emscripten-browser-file) (MIT)


CHANGELOG.txt
v0.20.0
* BREAK instruction
* F5 to pause/continue simulation

v0.19.0
* High Contrast palette
* Display/Recording Orientation setting

v0.18.0
* PROGMEM space window
* Source window
* Locals window

v0.17.0
* Add auto-breaks for a variety of runtime errors

v0.16.0
* Record audio to WAV

v0.15.0
* Add auto save of EEPROM and saved FX sectors
* Player and debugger now share same logic

v0.14.0
* Display scan lines setting
* Low contrast palette

v0.13.0
* HQx family of filters

v0.12.0
* Palette support in display and recordings

v0.11.0
* Sound window (waveform and spectrum views)

v0.10.0
* Serial monitor

v0.9.0
* Dedicated settings modal dialog
* Display and recording filtering with ScaleNx family of filters

v0.8.0
* Use @unwiredben's CORS proxy to enable loading files via URL parameters:
  - g=[1,2,3] to select grayscale mode (0 for monochrome, 1 for 3-level, 2 for 4-level)
  - z=[0,1] to control full zoom mode
  - hex=<url> to load .hex file
  - bin=<url> to load .bin file
  - arduboy=<url> to load .arduboy file
* "Full Zoom" mode: press F to toggle between full display and debugger

v0.7.0
* Snapshot support:
  - Press F4 to save a .snapshot file
  - Snapshots contain full simulation state and debug info
  - Open .snapshot files to resume

v0.6.0
* Add LED emulation and LEDs window
* Fix bug in sbc instruction

v0.5.0
* Add PNG screenshot and GIF recording (not available in Web yet):
  - F1 to take PNG screenshot of entire app window
  - F2 to take PNG screenshot of running Arduboy game
  - F3 to toggle GIF recording of running Arduboy game

v0.4.0
* CPU Usage window

v0.3.0
* Version displays in main menu bar
* EEPROM window

v0.2.0
* Fixes to SPI and FX emulation. Thanks Mr.Blinky!
* Display Internals window shows internal RAM pointer.

v0.1.0
* Added "FX Data" and "FX Internals" windows
* Fixed bug with SPIF timing

v0.0.0
* First released version.


src/common.cpp
#include "common.hpp"

#include <fstream>
#include <strstream>
#include <algorithm>

#include <string.h>
#include <stdlib.h>
#include <math.h>

#include <fmt/format.h>

#include <imgui.h>

#ifndef ARDENS_NO_DEBUGGER
#include <implot.h>
#endif

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#include <emscripten/html5.h>
#include "emscripten-browser-file/emscripten_browser_file.h"
#endif

#if !defined(__EMSCRIPTEN__)
#define ALLOW_SCREENSHOTS 1
#else
#define ALLOW_SCREENSHOTS 1
#endif

#if ALLOW_SCREENSHOTS
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#endif

#ifdef ARDENS_NO_DEBUGGER
void process_sound_samples() {}
#endif

#ifndef ARDENS_NO_GUI
extern unsigned char const ProggyVector[198188];
#endif

texture_t display_texture{};
texture_t display_buffer_texture{};
std::unique_ptr<absim::arduboy_t> arduboy;

int profiler_selected_hotspot = -1;
int disassembly_scroll_addr = -1;
bool scroll_addr_to_top = false;
int simulation_slowdown = 1000;
int fx_data_scroll_addr = -1;

int display_buffer_addr = -1;
int display_buffer_w = 128;
int display_buffer_h = 64;

static ImGuiStyle default_style;

float pixel_ratio = 1.f;
std::string dropfile_err;
bool loading_indicator = false;
uint64_t ms_since_start;

uint64_t ms_since_touch;
bool first_touch = false;
std::unordered_map<uintptr_t, touch_point_t> touch_points;

bool done = false;
bool layout_done = false;
bool settings_loaded = false;
#ifdef __EMSCRIPTEN__
bool fs_ready = false;
#else
bool fs_ready = true;
#endif

#ifdef __EMSCRIPTEN__
void file_download(
    char const* fname,
    char const* download_fname,
    char const* mime_type)
{
    std::ifstream f(fname, std::ios::binary);
    std::vector<char> data(
        (std::istreambuf_iterator<char>(f)),
        std::istreambuf_iterator<char>());
    emscripten_browser_file::download(download_fname, mime_type, data.data(), data.size());
}
#endif

#ifndef ARDENS_PLATFORM_SDL
#ifdef _WIN32
#define NOMINMAX
#include <windows.h>
#include <shellapi.h>
#endif
void platform_open_url(char const *url)
{
#ifdef __EMSCRIPTEN__
    EM_ASM(
        { window.open(UTF8ToString($0), "_blank"); },
        url);
#endif
#ifdef _WIN32
    ShellExecuteA(0, NULL, url, NULL, NULL, SW_SHOWNORMAL);
#endif
#if defined(__APPLE__) || defined(__MACH__)
    (void)system(fmt::format("open {}", url).c_str());
#endif
#if defined(__linux__) || defined(__FreeBSD__)
    (void)system(fmt::format("xdg-open {}", url).c_str());
#endif
}
#endif

extern "C" int load_file(char const* filename, uint8_t const* data, size_t size)
{
    std::istrstream f((char const*)data, size);
    dropfile_err = arduboy->load_file(filename, f);
    if(dropfile_err.empty()) load_savedata();
    return 0;
}

extern "C" int setparam(char const* name, char const* value)
{
    if(!name || !value) return 0;
    int nvalue = atoi(value);
    bool bvalue = (nvalue != 0);
    int r = 0;
    if(!strcmp(name, "z"))
    {
        settings.fullzoom = bvalue;
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "g") || !strcmp(name, "grid"))
    {
        if(!strcmp(value, "none"))
            settings.display_pixel_grid = PGRID_NONE;
        else if(!strcmp(value, "normal"))
            settings.display_pixel_grid = PGRID_NORMAL;
        else if(!strcmp(value, "red"))
            settings.display_pixel_grid = PGRID_RED;
        else if(!strcmp(value, "green"))
            settings.display_pixel_grid = PGRID_GREEN;
        else if(!strcmp(value, "blue"))
            settings.display_pixel_grid = PGRID_BLUE;
        else if(!strcmp(value, "cyan"))
            settings.display_pixel_grid = PGRID_CYAN;
        else if(!strcmp(value, "magenta"))
            settings.display_pixel_grid = PGRID_MAGENTA;
        else if(!strcmp(value, "yellow"))
            settings.display_pixel_grid = PGRID_YELLOW;
        else if(!strcmp(value, "white"))
            settings.display_pixel_grid = PGRID_WHITE;
        else
            settings.display_pixel_grid = std::clamp<int>(nvalue, PGRID_MIN, PGRID_MAX);
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "p") || !strcmp(name, "palette"))
    {
        if(!strcmp(value, "default"))
            settings.display_palette = PALETTE_DEFAULT;
        else if(!strcmp(value, "retro"))
            settings.display_palette = PALETTE_RETRO;
        else if(!strcmp(value, "lowcontrast"))
            settings.display_palette = PALETTE_LOW_CONTRAST;
        else if(!strcmp(value, "highcontrast"))
            settings.display_palette = PALETTE_HIGH_CONTRAST;
        else
            settings.display_palette = std::clamp<int>(nvalue, PALETTE_MIN, PALETTE_MAX);
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "af") || !strcmp(name, "autofilter"))
    {
        settings.display_auto_filter = bvalue;
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "f") || !strcmp(name, "filter"))
    {
        if(!strcmp(value, "none"))
            settings.display_filtering = FILTER_NONE;
        else if(!strcmp(value, "scale2x"))
            settings.display_filtering = FILTER_SCALE2X;
        else if(!strcmp(value, "scale3x"))
            settings.display_filtering = FILTER_SCALE3X;
        else if(!strcmp(value, "scale4x"))
            settings.display_filtering = FILTER_SCALE4X;
        else if(!strcmp(value, "hq2x"))
            settings.display_filtering = FILTER_HQ2X;
        else if(!strcmp(value, "hq3x"))
            settings.display_filtering = FILTER_HQ3X;
        else if(!strcmp(value, "hq4x"))
            settings.display_filtering = FILTER_HQ4X;
        else
            settings.display_filtering = std::clamp<int>(nvalue, FILTER_MIN, FILTER_MAX);
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "ds") || !strcmp(name, "downsample"))
    {
        settings.display_downsample = std::clamp<int>(nvalue, 1, 4);
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "ori") || !strcmp(name, "orientation"))
    {
        if(!strcmp(value, "0") || !strcmp(value, "normal"))
            settings.display_orientation = 0;
        else if(!strcmp(value, "90") || !strcmp(value, "cw") || !strcmp(value, "cw90"))
            settings.display_orientation = 1;
        else if(!strcmp(value, "180") || !strcmp(value, "flip"))
            settings.display_orientation = 2;
        else if(!strcmp(value, "270") || !strcmp(value, "ccw") || !strcmp(value, "ccw90"))
            settings.display_orientation = 3;
        else
            settings.display_orientation = std::clamp<int>(nvalue, 0, 3);
        update_settings();
        r = 1;
    }
    else if(!strcmp(name, "loading"))
    {
        loading_indicator = bvalue;
        r = 1;
    }
    return r;
}

extern "C" void postsyncfs()
{
    fs_ready = true;
    load_savedata();
}

bool update_pixel_ratio()
{
    float ratio = platform_pixel_ratio();
    bool changed = (ratio != pixel_ratio);
    pixel_ratio = ratio;
    return changed;
}

void define_font()
{
    ImGuiIO& io = ImGui::GetIO();
    ImFontConfig cfg;
    cfg.FontDataOwnedByAtlas = false;
    cfg.RasterizerMultiply = 1.5f;
    cfg.OversampleH = 2;
    cfg.OversampleV = 2;
    io.Fonts->Clear();
#ifndef ARDENS_NO_GUI
    io.Fonts->AddFontFromMemoryTTF(
        (void*)ProggyVector, sizeof(ProggyVector), 13.f * pixel_ratio, &cfg);
#endif
}

void rebuild_fonts()
{
    platform_destroy_fonts_texture();
    define_font();
    platform_create_fonts_texture();
}

void rescale_style()
{
    auto& style = ImGui::GetStyle();
    style = default_style;
    style.ScaleAllSizes(pixel_ratio);
}

void shutdown()
{
#ifndef ARDENS_NO_DEBUGGER
    ImPlot::DestroyContext();
#endif
}

void init()
{
    printf(
        "Ardens "
#ifdef ARDENS_PLAYER
        "Player "
#endif
        ARDENS_VERSION " by Peter Brown\n");

#ifdef __EMSCRIPTEN__
    EM_ASM(
        FS.mkdir('/offline');
    FS.mount(IDBFS, {}, '/offline');
    FS.syncfs(true, function(err) { ccall('postsyncfs', 'v'); });
    );
#endif

    arduboy = std::make_unique<absim::arduboy_t>();

#ifndef ARDENS_NO_DEBUGGER
    ImPlot::CreateContext();
#endif

#ifndef ARDENS_NO_SAVED_SETTINGS
    init_settings();
#endif

    ImGuiIO& io = ImGui::GetIO();
#if !defined(__EMSCRIPTEN__) && !defined(ARDENS_NO_SAVED_SETTINGS)
    ImGui::LoadIniSettingsFromDisk(io.IniFilename);
    settings_loaded = true;
#endif

    io.ConfigFlags |= ImGuiConfigFlags_DockingEnable;

#if defined(__EMSCRIPTEN__) || defined(ARDENS_NO_SAVED_SETTINGS)
    io.IniFilename = nullptr;
#endif

    ImGui::StyleColorsDark();
    ImGui::GetStyle().Colors[ImGuiCol_WindowBg] = ImVec4(0.06f, 0.06f, 0.06f, 1.00f);
    ImGuiStyle& style = ImGui::GetStyle();
    style.WindowRounding = 0.0f;
    style.Colors[ImGuiCol_WindowBg].w = 1.0f;
    style.Colors[ImGuiCol_PopupBg].w = 1.0f;

    default_style = style;

    arduboy->fx.erase_all_data();
    arduboy->reset();
    arduboy->fx.min_page = 0xffff;
    arduboy->fx.max_page = 0xffff;

    ms_since_start = 0;
    ms_since_touch = MS_SHOW_TOUCH_CONTROLS;
}

void save_screenshot()
{
    char fname[256];
    time_t rawtime;
    struct tm* ti;
    time(&rawtime);
    ti = localtime(&rawtime);
    (void)snprintf(fname, sizeof(fname),
        "screenshot_%04d%02d%02d%02d%02d%02d.png",
        ti->tm_year + 1900, ti->tm_mon + 1, ti->tm_mday,
        ti->tm_hour + 1, ti->tm_min, ti->tm_sec);
    int z = recording_filter_zoom();
    int w = 128 * z;
    int h = 64 * z;
    int stride = 128 * 4 * z;
    if(settings.recording_orientation & 1)
    {
        stride /= 2;
        std::swap(w, h);
    }
#ifdef __EMSCRIPTEN__
    stbi_write_png("screenshot.png", w, h, 4, recording_pixels(true), stride);
    file_download("screenshot.png", fname, "image/x-png");
#else
    stbi_write_png(fname, w, h, 4, recording_pixels(true), stride);
#endif
}

void toggle_recording()
{
    screen_recording_toggle(recording_pixels(false));
    if(wav_recording || settings.record_wav)
        wav_recording_toggle();
}

void take_snapshot()
{
#ifndef ARDENS_NO_SNAPSHOTS
    char fname[256];
    time_t rawtime;
    struct tm* ti;
    time(&rawtime);
    ti = localtime(&rawtime);
    (void)snprintf(fname, sizeof(fname),
        "ardens_%04d%02d%02d%02d%02d%02d.snapshot",
        ti->tm_year + 1900, ti->tm_mon + 1, ti->tm_mday,
        ti->tm_hour + 1, ti->tm_min, ti->tm_sec);
#ifdef __EMSCRIPTEN__
    std::ofstream f("ardens.snapshot", std::ios::binary);
    if(arduboy->save_snapshot(f))
    {
        f.close();
        file_download("ardens.snapshot", fname, "application/octet-stream");
    }
#else
    std::ofstream f(fname, std::ios::binary);
    arduboy->save_snapshot(f);
#endif
#endif
}

void frame_logic()
{
    ImGuiIO& io = ImGui::GetIO();

    if(!touch_points.empty())
        ms_since_touch = 0;

#ifdef __EMSCRIPTEN__
    if(done) emscripten_cancel_main_loop();
#endif

    // advance simulation
    uint64_t dt = platform_get_ms_dt();
    ms_since_start += dt;
    ms_since_touch += dt;
    if(dt > 30) dt = 30;
    if(!arduboy->paused)
    {
        // PINF: 4,5,6,7=D,L,R,U
        // PINE: 6=A
        // PINB: 4=B
        uint8_t pinf = 0xf0;
        uint8_t pine = 0x40;
        uint8_t pinb = 0x10;

        if(!ImGui::GetIO().WantCaptureKeyboard)
        {
            std::array<ImGuiKey, 4> keys =
            {
                ImGuiKey_UpArrow,
                ImGuiKey_RightArrow,
                ImGuiKey_DownArrow,
                ImGuiKey_LeftArrow,
            };
            std::array<int, 4> tkeys =
            {
                TOUCH_U, TOUCH_R, TOUCH_D, TOUCH_L,
            };
            auto touch = touched_buttons();

            std::rotate(keys.begin(), keys.begin() + settings.display_orientation, keys.end());
            std::rotate(tkeys.begin(), tkeys.begin() + settings.display_orientation, tkeys.end());

            if(ImGui::IsKeyDown(keys[0]) || touch.btns[tkeys[0]]) pinf &= ~0x80;
            if(ImGui::IsKeyDown(keys[1]) || touch.btns[tkeys[1]]) pinf &= ~0x40;
            if(ImGui::IsKeyDown(keys[2]) || touch.btns[tkeys[2]]) pinf &= ~0x10;
            if(ImGui::IsKeyDown(keys[3]) || touch.btns[tkeys[3]]) pinf &= ~0x20;

            if(ImGui::IsKeyDown(ImGuiKey_A) || touch.btns[TOUCH_A]) pine &= ~0x40;
            if(ImGui::IsKeyDown(ImGuiKey_B) || ImGui::IsKeyDown(ImGuiKey_S) || touch.btns[TOUCH_B]) pinb &= ~0x10;

            arduboy->cpu.data[0x23] = pinb;
            arduboy->cpu.data[0x2c] = pine;
            arduboy->cpu.data[0x2f] = pinf;
        }

#if PROFILING
        dt = 200;
#endif

        bool prev_paused = arduboy->paused;
        arduboy->frame_bytes_total = 1024;

        arduboy->cpu.enabled_autobreaks.reset();
#ifndef ARDENS_NO_GUI
        arduboy->cpu.enabled_autobreaks.set(absim::AB_BREAK);
        for(int i = 1; i < absim::AB_NUM; ++i)
            if(settings.ab.index(i))
                arduboy->cpu.enabled_autobreaks.set(i);
#endif

        arduboy->allow_nonstep_breakpoints =
            arduboy->break_step == 0xffffffff || settings.enable_step_breaks;
        arduboy->display.enable_filter = settings.display_auto_filter;

        constexpr uint64_t MS_TO_PS = 1000000000ull;
        uint64_t dtps = dt * MS_TO_PS * 1000 / simulation_slowdown;
        if(gif_recording)
        {
            constexpr uint64_t DT_20_MS = 20 * MS_TO_PS;
            uint64_t ps = DT_20_MS - gif_ps_rem;
            while(dtps >= ps)
            {
                arduboy->advance(ps);
                send_gif_frame(2, recording_pixels(false));
                dtps -= ps;
                ps = DT_20_MS;
                gif_ps_rem = 0;
            }
            gif_ps_rem += dtps;
        }
        if(dtps > 0)
            arduboy->advance(dtps);

        check_save_savedata();

        if(arduboy->paused && !prev_paused)
            disassembly_scroll_addr = arduboy->cpu.pc * 2;
        //if(!settings.enable_stack_breaks)
        //    arduboy->cpu.stack_overflow = false;

        // consume sound buffer
        send_wav_audio();
        process_sound_samples();
#if !PROFILING
        if(!arduboy->cpu.sound_buffer.empty() && simulation_slowdown == 1000)
            platform_send_sound();
#endif
        arduboy->cpu.sound_buffer.clear();
    }
    else
    {
        arduboy->break_step = 0xffffffff;
    }

    // update framebuffer texture
    if(arduboy->cpu.decoded)
    {
        recreate_display_texture();
        int z = display_texture_zoom;
        std::vector<uint8_t> pixels;
        pixels.resize(128 * 64 * 4 * z * z);
        scalenx(pixels.data(), arduboy->display.filtered_pixels.data(), true);
        platform_update_texture(display_texture, pixels.data(), pixels.size());

#if ALLOW_SCREENSHOTS
        if(arduboy->cpu.decoded && ImGui::IsKeyPressed(ImGuiKey_F4, false))
            take_snapshot();
        if(arduboy->cpu.decoded && ImGui::IsKeyPressed(ImGuiKey_F2, false))
            save_screenshot();
        if(arduboy->cpu.decoded && ImGui::IsKeyPressed(ImGuiKey_F3, false))
            toggle_recording();
#endif
    }

    if(!ImGui::GetIO().WantCaptureKeyboard && ImGui::IsKeyPressed(ImGuiKey_P, false))
    {
        settings.fullzoom = !settings.fullzoom;
        update_settings();
    }

#ifndef ARDENS_NO_DEBUGGER
    if(ImGui::IsKeyPressed(ImGuiKey_F5, false))
        arduboy->paused = !arduboy->paused;
#endif

    if(ImGui::IsKeyPressed(ImGuiKey_F11, false))
        platform_toggle_fullscreen();

    if(!gif_recording && ImGui::IsKeyPressed(ImGuiKey_R, false))
    {
        settings.display_orientation = (settings.display_orientation + 1) % 4;
        update_settings();
    }

    if(update_pixel_ratio())
    {
        rescale_style();
        rebuild_fonts();
    }
}

void imgui_content()
{
    ImGuiIO& io = ImGui::GetIO();

    if(!arduboy->cpu.decoded)
    {
        auto* d = ImGui::GetBackgroundDrawList(ImGui::GetMainViewport());
        auto size = ImGui::GetMainViewport()->Size;
        float w = 100.f * pixel_ratio;
        float pad = 25.f * pixel_ratio;
        w = std::min(w, size.x - pad);
        w = std::min(w, size.y - pad);
        w = std::max(w, pad);
        float w2 = w * 0.5f;
        float cx = size.x * 0.5f;
        float cy = size.y * 0.5f;
        constexpr uint8_t shade = 200;
        auto color = IM_COL32(shade, shade, shade, 255);
        float thickness = 4.f * pixel_ratio;
        if(loading_indicator)
        {
            for(int i = 0; i < 12; ++i)
            {
                float tx = cosf(3.1415926536f / 6 * i) * 0.5f;
                float ty = sinf(3.1415926536f / 6 * i) * 0.5f;
                int ti = (int)round(fmod(ImGui::GetTime(), 1.0) * 12);
                if(ti > 11) ti = 11;
                if(ti < 0) ti = 0;
                ti -= i;
                if(ti < 0) ti += 12;
                int tf = (int)roundf((float)(11 - ti - 6) * (255.f / 5));
                if(tf < 0) tf = 0;
                d->AddCircleFilled(
                    { cx + tx * w2, cy + ty * w2 },
                    thickness,
                    IM_COL32(shade, shade, shade, tf));
            }
        }
        else
        {
            d->AddRect(
                { cx - w2, cy - w2 },
                { cx + w2, cy + w2 },
                color,
                10.f * pixel_ratio,
                0,
                thickness);
            d->AddLine(
                { cx, cy - w2 * 0.5f },
                { cx, cy },
                color,
                thickness);
            d->AddTriangle(
                { cx - w2 * 0.5f, cy },
                { cx + w2 * 0.5f, cy },
                { cx, cy + w2 * 0.5f },
                color,
                thickness);
        }
    }

#ifndef ARDENS_NO_DEBUGGER
    if(settings.fullzoom)
        view_player();
    else
        view_debugger();
#else
    view_player();
#endif

#ifndef ARDENS_NO_GUI
    modal_settings();
#endif

#ifndef ARDENS_NO_GUI
    if(!dropfile_err.empty() && !ImGui::IsPopupOpen("File Load Error"))
        ImGui::OpenPopup("File Load Error");

    ImGui::SetNextWindowSize({ 500 * pixel_ratio, 0 });
    if(ImGui::BeginPopupModal("File Load Error", NULL,
        ImGuiWindowFlags_AlwaysAutoResize |
        ImGuiWindowFlags_NoSavedSettings))
    {
        ImGui::PushTextWrapPos(0.0f);
        ImGui::TextUnformatted(dropfile_err.c_str());
        ImGui::PopTextWrapPos();
        if(ImGui::Button("OK", ImVec2(120 * pixel_ratio, 0)) || ImGui::IsKeyPressed(ImGuiKey_Enter))
        {
            dropfile_err.clear();
            ImGui::CloseCurrentPopup();
        }
        ImGui::EndPopup();
    }
#endif

#ifndef ARDENS_NO_GUI
    if(arduboy->cpu.should_autobreak_gui())
        ImGui::OpenPopup("Auto-Break");

    static std::array<char const*, absim::AB_NUM> const AB_REASONS =
    {
        "BREAK Instruction",
        "Stack Overflow",
        "Null Dereference: a load/store was executed at address 0x0000.",
        "Null-Relative Dereference: a load/store with displacement (ldd/std) was executed with a null base pointer.",
        "Out-of-bounds Dereference: a load/store was executed at an address outside of RAM bounds.",
        "EEPROM Out-of-bounds: EEPROM was accessed at an address greater than 0x3ff.",
        "Out-of-bounds Indirect Jump: an indirect jump (ijmp) was executed to an address outside of program memory.",
        "Out-of-bounds PC: the PC is now pointing past the last byte of program memory.",
        "Unknown Instruction: encountered an instruction that is invalid or currently unsupported.",
        "SPI Write Collision: attempted to write to SPI data register before SPI was ready.",
        "FX Busy: attempted to execute FX command while flash chip was busy."
    };

    ImGui::SetNextWindowSize({ 300 * pixel_ratio, 0 });
    if(ImGui::BeginPopupModal("Auto-Break", NULL,
        ImGuiWindowFlags_AlwaysAutoResize |
        ImGuiWindowFlags_NoSavedSettings))
    {
        int ab = 0;
        auto mask = arduboy->cpu.autobreaks & arduboy->cpu.enabled_autobreaks;
        for(size_t i = 1; i < mask.size(); ++i)
        {
            if(mask.test(i))
            {
                ab = i;
                break;
            }
        }

        ImGui::PushTextWrapPos(0.0f);
        ImGui::TextUnformatted("A runtime error was caught.");
        ImGui::Separator();
        ImGui::TextUnformatted(AB_REASONS[ab]);
        ImGui::PopTextWrapPos();
        if(ImGui::Button("OK", ImVec2(120 * pixel_ratio, 0)) || ImGui::IsKeyPressed(ImGuiKey_Enter))
        {
            arduboy->cpu.autobreaks = 0;
            ImGui::CloseCurrentPopup();
        }
        ImGui::EndPopup();
    }
#endif

#ifndef ARDENS_NO_GUI
    modal_about();
#endif

#ifndef ARDENS_NO_GUI
#if PROFILING
    ImGui::ShowMetricsWindow();
#endif
    //ImGui::ShowDemoWindow();
#endif
}
